### PR TITLE
De-virtualize compareAt and getUInt calls in comparator for LowCardinality column

### PR DESCRIPTION
**Pull request by davenger**
### Changelog category (leave one):
- Performance Improvement

### Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):
Speedup sorting of `LowCardinality(String)` columns by de-virtualizing calls in comparator

### Documentation entry for user-facing changes

- [ ] Documentation is written (mandatory for new features)

> Information about CI checks: https://clickhouse.com/docs/en/development/continuous-integration/

#### CI Settings (Only check the boxes if you know what you are doing):
- [ ]  Allow: All Required Checks
- [ ]  Allow: Stateless tests
- [ ]  Allow: Stateful tests
- [ ]  Allow: Integration Tests
- [ ]  Allow: Performance tests
- [ ]  Allow: All Builds
- [ ]  Allow: batch 1, 2 for multi-batch jobs
- [ ]  Allow: batch 3, 4, 5, 6 for multi-batch jobs
---
- [ ]  Exclude: Style check
- [ ]  Exclude: Fast test
- [ ]  Exclude: All with ASAN
- [ ]  Exclude: All with TSAN, MSAN, UBSAN, Coverage
- [ ]  Exclude: All with aarch64, release, debug
---
- [ ]  Run only fuzzers related jobs (libFuzzer fuzzers, AST fuzzers, etc.)
- [ ]  Exclude: AST fuzzers
---
- [ ]  Do not test
- [ ]  Woolen Wolfdog
- [ ]  Upload binaries for special builds
- [ ]  Disable merge-commit
- [ ]  Disable CI cache

### CODE REVIEW

**Comment by robot-ch-test-poll1** [comment]
*This is an automated comment for commit 989c85ef52cadf5c1ddafa55ff76c0353438af0f with description of existing statuses. It's updated for the latest CI running*

[✅ Click here](https://s3.amazonaws.com/clickhouse-test-reports/72337/989c85ef52cadf5c1ddafa55ff76c0353438af0f/ci_running.html) to open a full report in a separate page

<details><summary>Successful checks</summary>
<table>
<thead><tr><th>Check name</th><th>Description</th><th>Status</th></tr></thead>
<tbody>
<tr><td>Docs check</td><td>Builds and tests the documentation</td><td>✅ success</td></tr>
<tbody>
</table>
</details>

File: src/Columns/ColumnLowCardinality.cpp [hunk 1]
```diff
@@ -426,29 +426,85 @@ void ColumnLowCardinality::getPermutation(IColumn::PermutationSortDirection dire
...
 }
 
-void ColumnLowCardinality::updatePermutation(IColumn::PermutationSortDirection direction, IColumn::PermutationSortStability stability,
-                                        size_t limit, int nan_direction_hint, IColumn::Permutation & res, EqualRanges & equal_ranges) const
+namespace
 {
-    bool ascending = direction == IColumn::PermutationSortDirection::Ascending;
 
-    auto comparator = [this, ascending, stability, nan_direction_hint](size_t lhs, size_t rhs)
+/// Compator for sorting LowCardinality column with the help of sorted dictionary.
```
Comment (davenger) [inline src/Columns/ColumnLowCardinality.cpp:432]:
```suggestion
/// Comparator for sorting LowCardinality column with the help of sorted dictionary.
```

**Comment by davenger** [comment]
https://s3.amazonaws.com/clickhouse-test-reports/72337/140f5987da4dbbce3ad082f5a84a501ef3d3e346/performance_comparison__release__[1_4]/report.html
image

**Comment by nickitat** [comment]
msan report should be fixed by https://github.com/ClickHouse/ClickHouse/pull/72226

File: src/Columns/ColumnLowCardinality.cpp [hunk 2]
```diff
@@ -426,29 +426,85 @@ void ColumnLowCardinality::getPermutation(IColumn::PermutationSortDirection dire
...
+{
+    /// Cast indexes column to the real type so that compareAt and getUInt methods can be inlined.
+    const IndexColumn * real_indexes = assert_cast<const IndexColumn *>(&getIndexes());
+
+    IColumn::Permutation dict_perm;
+    getDictionary().getNestedColumn()->getPermutation(direction, stability, 0, nan_direction_hint, dict_perm);
+
+    PaddedPODArray<UInt64> position_by_index(dict_perm.size());
+    for (size_t i = 0; i < dict_perm.size(); ++i)
+        position_by_index[dict_perm[i]] = i;
```
Comment (antonio2368) [inline src/Columns/ColumnLowCardinality.cpp:474]: not sure how smart is compiler but maybe create `dict_perm` and `position_by_index` in `updatePermutation` so same code is not generated for each type?

Comment (davenger) [reply src/Columns/ColumnLowCardinality.cpp]: moved it to `updatePermutation` as you suggested

File: src/Columns/ColumnLowCardinality.cpp [hunk 3]
```diff
@@ -426,29 +426,91 @@ void ColumnLowCardinality::getPermutation(IColumn::PermutationSortDirection dire
...
+{
+    IColumn::Permutation dict_perm;
+    getDictionary().getNestedColumn()->getPermutation(direction, stability, 0, nan_direction_hint, dict_perm);
+
+    /// This is a paranoid check, but in other places in code empty permutation is used to indicate that no sorting is needed.
+    if (dict_perm.size() != getDictionary().size())
+        throw Exception(ErrorCodes::LOGICAL_ERROR,
+            "Dictionary permutation size {} is equal to dictionary size {}. It is a bug.",
+            dict_perm.size(), getDictionary().size());
+
```
Comment (davenger) [inline src/Columns/ColumnLowCardinality.cpp:492]:
```suggestion

```

**Review by alesapin — APPROVED** [review]

**Pull request merged**